File: apps/dawnchat-android/android/app/src/main/java/com/dawnchat/app/auth/MobileAuthRepository.kt
package com.dawnchat.app.auth

import android.content.Context
import android.net.Uri
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.dawnchat.app.DawnChatApp
import com.dawnchat.app.R
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.status.SessionStatus
import io.github.jan.supabase.auth.user.UserInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit

data class PendingAuthState(
    val state: String,
    val deviceId: String,
    val createdAt: Long,
    val nextPath: String?,
)

/** UI-facing snapshot from Supabase Auth [UserInfo]; avoids binding views to SDK types. */
data class UserProfileSnapshot(
    val id: String,
    val email: String?,
    val phone: String?,
    val displayName: String?,
    val avatarUrl: String?,
    val createdAtIso: String?,
    val lastSignInAtIso: String?,
)

private fun JsonObject?.metadataString(vararg keys: String): String? {
    if (this == null) return null
    for (key in keys) {
        val prim = get(key) as? JsonPrimitive ?: continue
        val s = prim.contentOrNull?.trim()?.takeIf { it.isNotEmpty() } ?: continue
        return s
    }
    return null
}

private fun UserInfo.toProfileSnapshot(): UserProfileSnapshot {
    val meta = userMetadata
    val display = meta.metadataString("full_name", "name", "display_name", "nickname")
    val avatar = meta.metadataString("avatar_url", "picture", "avatar")
    return UserProfileSnapshot(
        id = id,
        email = email?.trim()?.takeIf { it.isNotEmpty() },
        phone = phone?.trim()?.takeIf { it.isNotEmpty() },
        displayName = display,
        avatarUrl = avatar,
        createdAtIso = createdAt?.toString(),
        lastSignInAtIso = lastSignInAt?.toString(),
    )
}

class MobileAuthRepository(private val context: Context) {

    private val prefs by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            PREFS_FILE,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    fun redirectUri(): String =
        context.getString(R.string.dawnchat_mobile_auth_redirect_uri).trim()

    fun bridgeBaseUrl(): String =
        context.getString(R.string.dawnchat_auth_bridge_base_url).trim()

    fun supabaseUrl(): String =
        context.getString(R.string.dawnchat_supabase_url).trim()

    fun supabaseAnonKey(): String =
        context.getString(R.string.dawnchat_supabase_anon_key).trim()

    fun isConfigReady(): Boolean =
        supabaseUrl().isNotBlank() && supabaseAnonKey().isNotBlank()

    fun matchesAuthRedirect(uri: Uri): Boolean =
        isAuthCallbackUri(uri, redirectUri())

    fun getOrCreateDeviceId(): String {
        val existing = prefs.getString(KEY_DEVICE_ID, null)
        if (!existing.isNullOrBlank()) return existing
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(KEY_DEVICE_ID, id).apply()
        return id
    }

    /**
     * Best-effort sync check (after storage load). Prefer [awaitSessionVerifiedForLaunch] on cold start.
     */
    fun isLoggedIn(): Boolean {
        val app = context.applicationContext as? DawnChatApp ?: return false
        val client = app.supabaseClientOrNull() ?: return false
        return client.auth.currentSessionOrNull() != null
    }

    suspend fun clearSession() = withContext(Dispatchers.IO) {
        val app = context.applicationContext as? DawnChatApp
        try {
            app?.supabaseClientOrNull()?.auth?.signOut()
        } catch (_: Exception) {
        }
        clearLegacyTokenKeys()
    }

    /**
     * Maps current Supabase Auth user to a stable snapshot for Mine / profile screens.
     * Uses in-memory session first; if session exists but user is missing, calls [retrieveUserForCurrentSession] once.
     */
    suspend fun loadProfileForUi(): UserProfileSnapshot? = withContext(Dispatchers.IO) {
        val app = context.applicationContext as? DawnChatApp ?: return@withContext null
        val client = app.supabaseClientOrNull() ?: return@withContext null
        val auth = client.auth
        var user: UserInfo? = auth.currentSessionOrNull()?.user
        if (user == null && auth.currentSessionOrNull() != null) {
            user = try {
                auth.retrieveUserForCurrentSession(updateSession = false)
            } catch (_: Exception) {
                null
            }
        }
        user?.toProfileSnapshot()
    }

    /**
     * Waits for Auth to leave [SessionStatus.Initializing], then validates session when network is available.
     * When offline, trusts the SDK-restored session if status is [SessionStatus.Authenticated].
     *
     * Manual QA: cold start with valid session; airplane mode + Authenticated; refresh failure → login;
     * Release build with `minifyEnabled true` if enabled later (see proguard-rules.pro).
     */
    suspend fun awaitSessionVerifiedForLaunch(): Boolean = withContext(Dispatchers.IO) {
        val app = context.applicationContext as? DawnChatApp ?: return@withContext false
        val client = app.supabaseClientOrNull() ?: return@withContext false
        if (!isConfigReady()) return@withContext false

        val status = client.auth.sessionStatus.first { it !is SessionStatus.Initializing }
        when (status) {
            is SessionStatus.Authenticated -> {
                if (!context.hasInternetCapability()) {
                    return@withContext true
                }
                try {
                    client.auth.retrieveUserForCurrentSession(updateSession = true)
                    true
                } catch (_: Exception) {
                    try {
                        client.auth.signOut()
                    } catch (_: Exception) {
                    }
                    clearLegacyTokenKeys()
                    false
                }
            }
            is SessionStatus.NotAuthenticated,
            is SessionStatus.RefreshFailure,
            -> false
            else -> false
        }
    }

    private fun clearLegacyTokenKeys() {
        prefs.edit()
            .remove(LEGACY_KEY_ACCESS)
            .remove(LEGACY_KEY_REFRESH)
            .remove(LEGACY_KEY_EXPIRES_AT)
            .apply()
    }

    fun writePending(pending: PendingAuthState) {
        val o = JSONObject()
        o.put("state", pending.state)
        o.put("deviceId", pending.deviceId)
        o.put("createdAt", pending.createdAt)
        if (pending.nextPath != null) o.put("nextPath", pending.nextPath)
        prefs.edit().putString(KEY_PENDING, o.toString()).apply()
    }

    fun readPending(): PendingAuthState? {
        val raw = prefs.getString(KEY_PENDING, null) ?: return null
        return try {
            val o = JSONObject(raw)
            val nextPath = if (o.has("nextPath")) o.optString("nextPath", "") else ""
            PendingAuthState(
                state = o.getString("state"),
                deviceId = o.getString("deviceId"),
                createdAt = o.getLong("createdAt"),
                nextPath = nextPath.takeIf { it.isNotEmpty() },
            )
        } catch (_: Exception) {
            clearPending()
            null
        }
    }

    fun clearPending() {
        prefs.edit().remove(KEY_PENDING).apply()
    }

    fun pendingValid(): PendingAuthState? {
        val p = readPending() ?: return null
        if (System.currentTimeMillis() - p.createdAt > DesktopAuthBridgeConstants.PENDING_TTL_MS) {
            clearPending()
            return null
        }
        return p
    }

    fun prepareBridgeLogin(nextPath: String = DesktopAuthBridgeConstants.DEFAULT_NEXT): String {
        val state = generateDesktopAuthState()
        val deviceId = getOrCreateDeviceId()
        writePending(
            PendingAuthState(
                state = state,
                deviceId = deviceId,
                createdAt = System.currentTimeMillis(),
                nextPath = nextPath,
            ),
        )
        return buildDesktopAuthBridgeUrl(
            baseUrl = bridgeBaseUrl(),
            state = state,
            deviceId = deviceId,
            redirectUri = redirectUri(),
            nextPath = nextPath,
            provider = null,
        )
    }

    suspend fun completeAuthFromCallback(uri: Uri): Result<Unit> = withContext(Dispatchers.IO) {
        if (!matchesAuthRedirect(uri)) {
            return@withContext Result.failure(IllegalArgumentException("not_auth_callback"))
        }
        val parsed = parseDesktopAuthCallback(uri)
        if (!parsed.error.isNullOrBlank()) {
            clearPending()
            return@withContext Result.failure(Exception(parsed.error))
        }
        val ticket = parsed.ticket?.trim().orEmpty()
        val state = parsed.state?.trim().orEmpty()
        if (ticket.isEmpty() || state.isEmpty()) {
            return@withContext Result.failure(Exception("登录回调缺少 ticket 或 state"))
        }
        exchangeTicket(ticket, state)
    }

    private suspend fun exchangeTicket(ticket: String, state: String): Result<Unit> {
        val pending = pendingValid()
            ?: return Result.failure(IllegalStateException("登录状态已失效，请重新发起登录"))
        if (pending.state != state) {
            clearPending()
            return Result.failure(IllegalStateException("登录状态校验失败，请重新登录"))
        }
        if (!isConfigReady()) {
            return Result.failure(IllegalStateException("请在 auth_config.xml 或 debug 覆盖中配置 Supabase URL 与 anon key"))
        }
        val app = context.applicationContext as? DawnChatApp
        val client = app?.supabaseClientOrNull()
            ?: return Result.failure(IllegalStateException("Supabase 客户端未初始化"))
        val exchanged = exchangeDesktopTicket(
            httpClient,
            supabaseUrl(),
            supabaseAnonKey(),
            ticket,
            state,
            pending.deviceId,
        )
        val tokens = exchanged.getOrElse { return Result.failure(it) }
        return try {
            client.auth.importSession(userSessionFromExchangeTokens(tokens))
            clearPending()
            clearLegacyTokenKeys()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    companion object {
        const val PREFS_FILE = "dawnchat_mobile_auth"
        const val LEGACY_KEY_ACCESS = "supabase_access_token"
        const val LEGACY_KEY_REFRESH = "supabase_refresh_token"
        const val LEGACY_KEY_EXPIRES_AT = "supabase_expires_at"
        private const val KEY_DEVICE_ID = "device_install_id"
        private const val KEY_PENDING = "auth_pending_json"
    }
}

File: apps/dawnchat-ios/ios/App/App/MobileAuthRepository.swift
import Foundation
import Network
import Security
import Supabase

struct PendingAuthState: Codable {
    let state: String
    let deviceId: String
    let createdAt: Int64
    let nextPath: String?
}

/// UI-facing snapshot from Supabase Auth `User`; avoids binding views to SDK types.
struct UserProfileSnapshot: Sendable {
    let id: String
    let email: String?
    let phone: String?
    let displayName: String?
    let avatarUrl: String?
    let createdAtIso: String?
    let lastSignInAtIso: String?
}

private func metadataString(from meta: [String: AnyJSON], keys: String...) -> String? {
    for key in keys {
        guard let value = meta[key] else { continue }
        if case let .string(s) = value {
            let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
            if !t.isEmpty { return t }
        }
    }
    return nil
}

private let profileDateFormatter: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime]
    return f
}()

private func mapSupabaseUserToProfile(_ u: User) -> UserProfileSnapshot {
    let meta = u.userMetadata
    let created = profileDateFormatter.string(from: u.createdAt)
    let lastIn: String? = u.lastSignInAt.map { profileDateFormatter.string(from: $0) }
    return UserProfileSnapshot(
        id: u.id.uuidString,
        email: u.email?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
        phone: u.phone?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
        displayName: metadataString(from: meta, keys: "full_name", "name", "display_name", "nickname"),
        avatarUrl: metadataString(from: meta, keys: "avatar_url", "picture", "avatar"),
        createdAtIso: created,
        lastSignInAtIso: lastIn
    )
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

/// Reachability for optional server-side session validation (offline: trust SDK session).
private final class DawnChatReachability {
    static let shared = DawnChatReachability()
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "com.dawnchat.reachability")
    private(set) var isConnected = false
    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            self?.isConnected = path.status == .satisfied
        }
        monitor.start(queue: queue)
    }
}

@MainActor
final class MobileAuthRepository {
    static let shared = MobileAuthRepository()

    private let defaults = UserDefaults.standard
    private let pendingKey = "dawnchat_mobile_auth_pending_json"
    private let deviceKey = "dawnchat_device_install_id"

    private var supabaseClient: SupabaseClient?

    private init() {}

    var redirectURI: String {
        (Bundle.main.object(forInfoDictionaryKey: "DawnChatMobileAuthRedirectURI") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "com.dawnchat.app://auth/callback"
    }

    var bridgeBaseURL: String {
        (Bundle.main.object(forInfoDictionaryKey: "DawnChatAuthBridgeBaseURL") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "https://dawnchat.com/desktop-auth/bridge"
    }

    var supabaseURL: String {
        (Bundle.main.object(forInfoDictionaryKey: "DawnChatSupabaseURL") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }

    var supabaseAnonKey: String {
        (Bundle.main.object(forInfoDictionaryKey: "DawnChatSupabaseAnonKey") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }

    var isConfigReady: Bool {
        !supabaseURL.isEmpty && !supabaseAnonKey.isEmpty
    }

    /// Creates the Supabase client, migrates legacy Keychain tokens once, then starts auto-refresh.
    func bootstrapSupabaseIfNeeded() async {
        _ = DawnChatReachability.shared
        guard supabaseClient == nil else { return }
        guard isConfigReady, let url = URL(string: supabaseURL) else { return }
        supabaseClient = SupabaseClient(supabaseURL: url, supabaseKey: supabaseAnonKey)
        await importLegacyKeychainIfNeeded()
    }

    func startSupabaseAutoRefresh() async {
        await supabaseClient?.auth.startAutoRefresh()
    }

    func stopSupabaseAutoRefresh() async {
        await supabaseClient?.auth.stopAutoRefresh()
    }

    private func importLegacyKeychainIfNeeded() async {
        guard let client = supabaseClient else { return }
        if (try? await client.auth.session) != nil {
            Self.clearLegacyKeychainTokens()
            return
        }
        guard let access = KeychainHelper.load(key: KeychainKeys.access)?.trimmingCharacters(in: .whitespacesAndNewlines),
              let refresh = KeychainHelper.load(key: KeychainKeys.refresh)?.trimmingCharacters(in: .whitespacesAndNewlines),
              !access.isEmpty, !refresh.isEmpty else { return }
        do {
            try await client.auth.setSession(accessToken: access, refreshToken: refresh)
            Self.clearLegacyKeychainTokens()
        } catch {
        }
    }

    func getOrCreateDeviceId() -> String {
        if let existing = defaults.string(forKey: deviceKey), !existing.isEmpty {
            return existing
        }
        let id = UUID().uuidString
        defaults.set(id, forKey: deviceKey)
        return id
    }

    func isLoggedIn() -> Bool {
        supabaseClient?.auth.currentSession != nil
    }

    /// Maps current Supabase Auth user to a stable snapshot for Mine / profile screens.
    func loadProfileForUi() async -> UserProfileSnapshot? {
        guard let client = supabaseClient, isConfigReady else { return nil }
        do {
            let session = try await client.auth.session
            return mapSupabaseUserToProfile(session.user)
        } catch {
            do {
                let u = try await client.auth.user()
                return mapSupabaseUserToProfile(u)
            } catch {
                return nil
            }
        }
    }

    /// Bearer token for Supabase Edge Functions (e.g. dawn-tts). Caller must not expose to WebView.
    func dawnTtsAccessToken() async throws -> String {
        guard let client = supabaseClient, isConfigReady else {
            throw NSError(
                domain: "DawnTts",
                code: -1,
                userInfo: [
                    "dawnCode": "supabase_config_missing",
                    NSLocalizedDescriptionKey: "Supabase URL or anon key is not configured",
                ]
            )
        }
        let session = try await client.auth.session
        let token = session.accessToken.trimmingCharacters(in: .whitespacesAndNewlines)
        if token.isEmpty {
            throw NSError(
                domain: "DawnTts",
                code: -2,
                userInfo: [
                    "dawnCode": "not_authenticated",
                    NSLocalizedDescriptionKey: "Not signed in or session has no access token",
                ]
            )
        }
        return token
    }

    /// After storage restore, optionally validates with Auth server when online.
    func awaitSessionVerifiedForLaunch() async -> Bool {
        guard let client = supabaseClient, isConfigReady else { return false }
        do {
            _ = try await client.auth.session
        } catch {
            return false
        }
        if !DawnChatReachability.shared.isConnected {
            return true
        }
        do {
            _ = try await client.auth.user()
            return true
        } catch {
            try? await client.auth.signOut()
            Self.clearLegacyKeychainTokens()
            return false
        }
    }

    func clearSession() async {
        if let client = supabaseClient {
            try? await client.auth.signOut()
        }
        Self.clearLegacyKeychainTokens()
    }

    private static func clearLegacyKeychainTokens() {
        KeychainHelper.delete(key: KeychainKeys.access)
        KeychainHelper.delete(key: KeychainKeys.refresh)
        KeychainHelper.delete(key: KeychainKeys.expiresAt)
    }

    func writePending(_ pending: PendingAuthState) {
        if let data = try? JSONEncoder().encode(pending) {
            defaults.set(String(data: data, encoding: .utf8), forKey: pendingKey)
        }
    }

    func readPending() -> PendingAuthState? {
        guard let raw = defaults.string(forKey: pendingKey),
              let data = raw.data(using: .utf8),
              let p = try? JSONDecoder().decode(PendingAuthState.self, from: data) else { return nil }
        return p
    }

    func clearPending() {
        defaults.removeObject(forKey: pendingKey)
    }

    func pendingValid() -> PendingAuthState? {
        guard let p = readPending() else { return nil }
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        if now - p.createdAt > DesktopAuthBridgeConstants.pendingTtlMs {
            clearPending()
            return nil
        }
        return p
    }

    func prepareBridgeLogin(nextPath: String = DesktopAuthBridgeConstants.defaultNext) throws -> URL {
        let state = MobileAuthBridge.generateState()
        let deviceId = getOrCreateDeviceId()
        writePending(PendingAuthState(state: state, deviceId: deviceId, createdAt: Int64(Date().timeIntervalSince1970 * 1000), nextPath: nextPath))
        return try MobileAuthBridge.buildBridgeURL(
            baseURL: bridgeBaseURL,
            state: state,
            deviceId: deviceId,
            redirectUri: redirectURI,
            nextPath: nextPath,
            provider: nil
        )
    }

    func matchesRedirect(_ url: URL) -> Bool {
        MobileAuthBridge.matchesRedirect(url, expectedRedirect: redirectURI)
    }

    func completeAuth(fromCallback url: URL) async throws {
        guard matchesRedirect(url) else {
            throw NSError(domain: "MobileAuth", code: 10, userInfo: [NSLocalizedDescriptionKey: "not_auth_callback"])
        }
        let parsed = MobileAuthBridge.parseCallback(url)
        if let err = parsed.error, !err.isEmpty {
            clearPending()
            throw NSError(domain: "MobileAuth", code: 11, userInfo: [NSLocalizedDescriptionKey: err])
        }
        guard let ticket = parsed.ticket?.trimmingCharacters(in: .whitespacesAndNewlines), !ticket.isEmpty,
              let state = parsed.state?.trimmingCharacters(in: .whitespacesAndNewlines), !state.isEmpty else {
            throw NSError(domain: "MobileAuth", code: 12, userInfo: [NSLocalizedDescriptionKey: "登录回调缺少 ticket 或 state"])
        }
        try await exchangeTicket(ticket: ticket, state: state)
    }

    private func exchangeTicket(ticket: String, state: String) async throws {
        guard let pending = pendingValid() else {
            throw NSError(domain: "MobileAuth", code: 13, userInfo: [NSLocalizedDescriptionKey: "登录状态已失效，请重新发起登录"])
        }
        if pending.state != state {
            clearPending()
            throw NSError(domain: "MobileAuth", code: 14, userInfo: [NSLocalizedDescriptionKey: "登录状态校验失败，请重新登录"])
        }
        guard isConfigReady else {
            throw NSError(domain: "MobileAuth", code: 15, userInfo: [NSLocalizedDescriptionKey: "请在 Info.plist 填写 DawnChatSupabaseURL 与 DawnChatSupabaseAnonKey"])
        }
        guard let client = supabaseClient else {
            throw NSError(domain: "MobileAuth", code: 19, userInfo: [NSLocalizedDescriptionKey: "Supabase 客户端未初始化"])
        }
        let urlStr = supabaseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/functions/v1/exchange-desktop-ticket"
        guard let reqUrl = URL(string: urlStr) else { throw NSError(domain: "MobileAuth", code: 16) }
        var req = URLRequest(url: reqUrl)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(supabaseAnonKey, forHTTPHeaderField: "apikey")
        let body: [String: Any] = [
            "desktop_ticket": ticket,
            "state": state,
            "device_id": pending.deviceId,
            "protocol_version": DesktopAuthBridgeConstants.protocolVersion,
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw NSError(domain: "MobileAuth", code: 17) }
        guard (200 ... 299).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String ?? text
            throw NSError(domain: "MobileAuth", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: msg.isEmpty ? "ticket 兑换失败" : msg])
        }
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let access = (json?["access_token"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let refresh = (json?["refresh_token"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if access.isEmpty || refresh.isEmpty {
            throw NSError(domain: "MobileAuth", code: 18, userInfo: [NSLocalizedDescriptionKey: "ticket 兑换结果缺少会话 token"])
        }
        try await client.auth.setSession(accessToken: access, refreshToken: refresh)
        Self.clearLegacyKeychainTokens()
        clearPending()
    }
}

private enum KeychainKeys {
    static let access = "dawnchat.supabase.access"
    static let refresh = "dawnchat.supabase.refresh"
    static let expiresAt = "dawnchat.supabase.expires_at"
}

private enum KeychainHelper {
    private static let service = "com.dawnchat.app.mobileauth"

    static func load(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var out: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &out)
        guard status == errSecSuccess, let d = out as? Data else { return nil }
        return String(data: d, encoding: .utf8)
    }

    static func delete(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
    }
}
